vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_statictensor.py

```diff
@@ -0,0 +1,140 @@
+                                     requires_grad=False)
+
+        weight_scale = Parameter(torch.empty(dim,
+                                             device="cuda",
+                                             dtype=torch.float32),
+                                 requires_grad=False)
+        weight_zero_point = Parameter(torch.empty(1,
+                                                  device="cuda",
+                                                  dtype=torch.int8),
+                                      requires_grad=False)
+
+        if not self.fake_quant:
+            params_dtype = torch.int8
+        weight = Parameter(torch.empty(sum(output_sizes_per_partition),
+                                       input_size_per_partition,
+                                       device="cuda",
+                                       dtype=params_dtype),
+                           requires_grad=False)
+
+        set_weight_attrs(weight, {"input_dim": 1, "output_dim": 0})
+        set_weight_attrs(
+            weight_scale, {
+                "shard_splitter": self.scales_shard_splitter,
+                "logical_widths": output_sizes_per_partition
+            })
+
+        # Register parameter with the layer; register weight loader with each parameter
+        layer.register_parameter("weight", weight)
+        set_weight_attrs(weight, {"weight_loader": weight_loader})
+        set_weight_attrs(weight,
+                         {"logical_widths": output_sizes_per_partition})
+
+        layer.register_parameter("input_scale", input_scale)
+        set_weight_attrs(input_scale, {"weight_loader": weight_loader})
+        layer.register_parameter("input_zero_point", input_zero_point)
+        set_weight_attrs(input_zero_point, {"weight_loader": weight_loader})
+        layer.register_parameter("weight_scale", weight_scale)
+        set_weight_attrs(weight_scale, {"weight_loader": weight_loader})
+        layer.register_parameter("weight_zero_point", weight_zero_point)
+        set_weight_attrs(weight_zero_point, {"weight_loader": weight_loader})
+
+    def apply_weights(self, layer: torch.nn.Module, x: torch.Tensor):
+        weight = layer.weight
+        weight_scale = layer.weight_scale
+        act_scale = layer.input_scale
+        logical_widths = weight.logical_widths
+
+        # Input quantize
+        #x_scales = torch.FloatTensor([act_scale[0].item()], device=torch.device("cpu"))
+        #x_q = self._quantize(x, x_scales, [x.shape[0]])
+        x_q = self._quantize_new(x, act_scale[0].item())
+
+        # Weight quantize
+        # TODO : try not to remove device-to-host copy. i.e. keep the non-duplicated version
+        # of scales in the CPU
+        if self.fake_quant:
+            w_scales = [
+                weight_scale[sum(logical_widths[:i])].item()
+                for i in range(len(logical_widths))
+            ]
+            w_scales = torch.FloatTensor(w_scales, device=torch.device("cpu"))
```

Comment:
why on the cpu?

Reply:
Here we are ultimately reading each value off the tensor and passing it by-value to the cuda kernels.